### PR TITLE
Changes to help sync cmcstl2 up with P0896R4

### PR DESCRIPTION
**Pull request by cjdb**
* Changes `safe_iterator_t`
   * Tests using `get_unsafe` revised or removed.
* copy[_*] changed to niebloids.
* Users of the copy family updated also (internally).

I'm not confident that I revised all of the tests correctly.

### CODE REVIEW

File: include/stl2/detail/algorithm/partial_sort_copy.hpp [hunk 1]
```diff
@@ -42,8 +42,8 @@ STL2_OPEN_NAMESPACE {
 	{
 		auto r = result_first;
 		if(r != result_last) {
-			std::tie(first, r) =
-				ext::copy(std::move(first), last, std::move(r), result_last);
+			auto [in, out] = ext::copy(std::move(first), last, std::move(r), result_last);
+			std::tie(first, r) = std::tie(in, out);
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/partial_sort_copy.hpp]: Ditto eww `tie` (I'll stop calling these out now and assume you're smart enough to grep for `tie`.)

File: include/stl2/detail/algorithm/rotate_copy.hpp [hunk 2]
```diff
@@ -21,32 +21,33 @@
...
-		return __stl2::rotate_copy(__stl2::begin(rng), std::forward<M>(middle),
-			__stl2::end(rng), std::forward<O>(result));
-	}
+	struct __rotate_copy_fn : private __niebloid {
+		template<ForwardIterator I, Sentinel<I> S, WeaklyIncrementable O>
+		requires IndirectlyCopyable<I, O>
+		constexpr rotate_copy_result<I, O>
+		operator()(I first, I middle, S last, O result) const
+		{
+			auto res = copy(middle, std::move(last), std::move(result));
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/rotate_copy.hpp]:
```suggestion
			auto res = __stl2::copy(middle, std::move(last), std::move(result));
```

File: include/stl2/detail/algorithm/rotate_copy.hpp [hunk 3]
```diff
@@ -21,32 +21,33 @@
...
+			res.out = copy(std::move(first), std::move(middle),
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/rotate_copy.hpp]:
```suggestion
			res.out = __stl2::copy(std::move(first), std::move(middle),
```

File: include/stl2/detail/algorithm/set_difference.hpp [hunk 4]
```diff
@@ -23,50 +23,53 @@
...
 					++first1;
+				} else {
+					if (!__stl2::invoke(comp, p2, p1)) {
+						++first1;
+					}
+					++first2;
 				}
-				++first2;
 			}
+			return copy(std::move(first1), std::move(last1), std::move(result));
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/set_difference.hpp]:
```suggestion
			return __stl2::copy(std::move(first1), std::move(last1), std::move(result));
```

File: include/stl2/detail/algorithm/set_difference.hpp [hunk 5]
```diff
@@ -23,50 +23,53 @@
...
+			class Comp = less, class Proj1 = identity, class Proj2 = identity>
+		requires Mergeable<iterator_t<R1>, iterator_t<R2>, O, Comp, Proj1, Proj2>
+		constexpr set_difference_result<safe_iterator_t<R1>, O>
+		operator()(R1&& r1, R2&& r2, O result,
+			Comp comp = {}, Proj1 proj1 = {}, Proj2 proj2 = {}) const
+		{
+			return (*this)(
+				begin(r1), end(r1),
+				begin(r2), end(r2),
+				std::forward<O>(result),
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/set_difference.hpp]:
```suggestion
				std::move(result),
```

File: include/stl2/detail/algorithm/set_symmetric_difference.hpp [hunk 6]
```diff
@@ -38,13 +38,13 @@ STL2_OPEN_NAMESPACE {
 	{
 		while (true) {
 			if (first1 == last1) {
-				std::tie(first2, result) =
-					__stl2::copy(std::move(first2), std::move(last2), std::move(result));
+				auto [in, out] = copy(std::move(first2), std::move(last2), std::move(result));
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/set_symmetric_difference.hpp]:
```suggestion
				auto [in, out] = __stl2::copy(std::move(first2), std::move(last2), std::move(result));
```

File: include/stl2/detail/algorithm/set_symmetric_difference.hpp [hunk 7]
```diff
@@ -38,13 +38,13 @@ STL2_OPEN_NAMESPACE {
...
+				std::tie(first2, result) = std::tie(in, out);
 				break;
 			}
 			if (first2 == last2) {
-				std::tie(first1, result) =
-					__stl2::copy(std::move(first1), std::move(last1), std::move(result));
+				auto [in, out] = copy(std::move(first1), std::move(last1), std::move(result));
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/set_symmetric_difference.hpp]:
```suggestion
				auto [in, out] = __stl2::copy(std::move(first1), std::move(last1), std::move(result));
```

File: include/stl2/detail/algorithm/set_union.hpp [hunk 8]
```diff
@@ -36,12 +36,12 @@ STL2_OPEN_NAMESPACE {
 	{
 		while (true) {
 			if (first1 == last1) {
-				auto res = __stl2::copy(std::move(first2), std::move(last2), std::move(result));
-				return {std::move(first1), std::move(res.in()), std::move(res.out())};
+				auto res = copy(std::move(first2), std::move(last2), std::move(result));
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/set_union.hpp]:
```suggestion
				auto res = __stl2::copy(std::move(first2), std::move(last2), std::move(result));
```

File: include/stl2/detail/algorithm/set_union.hpp [hunk 9]
```diff
@@ -36,12 +36,12 @@ STL2_OPEN_NAMESPACE {
...
+				return {std::move(first1), std::move(res.in), std::move(res.out)};
 			}
 			if (first2 == last2) {
-				auto res = __stl2::copy(std::move(first1), std::move(last1), std::move(result));
-				return {std::move(res.in()), std::move(first2), std::move(res.out())};
+				auto res = copy(std::move(first1), std::move(last1), std::move(result));
```
Comment (CaseyCarter) [inline include/stl2/detail/algorithm/set_union.hpp]:
```suggestion
				auto res = __stl2::copy(std::move(first1), std::move(last1), std::move(result));
```

File: include/stl2/detail/algorithm/copy_backward.hpp [hunk 10]
```diff
@@ -23,31 +23,32 @@
...
+				*--out = *--i;
+			}
+			return {std::move(last), std::move(out)};
+		}
+
+		template<BidirectionalRange R, BidirectionalIterator I>
+		constexpr copy_backward_result<safe_iterator_t<R>, I>
+		operator()(R&& r, I result) const
+		{
+			return (*this)(begin(r), end(r), std::forward<I>(result));
```
[the remark this answers is not in the document]
Comment (cjdb) [reply include/stl2/detail/algorithm/copy_backward.hpp]: Why was this resolved? (Similarly for the other s/`forward<T>`/`move`/ suggestions).

**Comment by CaseyCarter** [comment]
I rebased onto master, and your repo refuses me force-push access, so this PR is now #219.

**Pull request closed without merging**